# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [9]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [10]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
load_science_dataset_QL = pd.read_csv("science_data_large.csv")
display(load_science_dataset_QL.head(15))
display(load_science_dataset_QL.info())


,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


None

## 2. Splitting the dataset

In [11]:
# Take the pandas dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_features = load_science_dataset_QL.iloc[:,:-1]
Y_label = load_science_dataset_QL.iloc[:,-1]
X_set_train, X_set_test, Y_set_train, Y_set_test = train_test_split(X_features,\
                            Y_label,test_size=0.1,train_size= 0.9,random_state=42)
display(X_set_test)
display(Y_set_test)

,Temperature °C,Mols KCL
521,100,541
737,635,668
740,799,665
660,966,871
411,785,595
...,...,...
436,421,725
764,431,860
88,983,433
63,580,817


521    1.177623e+05
737    8.687293e+05
740    1.084893e+06
660    1.716039e+06
411    9.536850e+05
           ...     
436    6.305199e+05
764    7.676234e+05
88     8.684308e+05
63     9.737511e+05
826    4.786314e+03
Name: Size nm^3, Length: 100, dtype: float64

## 3. Perform a Linear Regression

In [12]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
dummi_data_sample = pd.DataFrame({"Temperature °C": [90], "Mols KCL" : [661]})
applied_Linear_Regression_model = LinearRegression()
applied_Linear_Regression_model.fit(X_set_train,Y_set_train)
linear_re_score = applied_Linear_Regression_model.score(X_set_test,Y_set_test)
predict_sample_point = applied_Linear_Regression_model.predict(dummi_data_sample).round(5)
coefficientd_values = applied_Linear_Regression_model.coef_.round(5)
intercept_value = applied_Linear_Regression_model.intercept_.round(5)
linear_re_data_frame = pd.DataFrame({
    "Sample" : [dummi_data_sample],
    'Predict sample output': [predict_sample_point],
    'Intercept point': [intercept_value],
    'coefficientd values': [coefficientd_values],
    'Linear Score': [linear_re_score],
})
display(linear_re_data_frame)

,Sample,Predict sample output,Intercept point,coefficientd values,Linear Score
0,Temperature °C Mols KCL 0 90 ...,[351173.13657],-409391.47958,"[866.14641, 1032.69507]",0.855247


Equation : $$h(x) = \hat{y}(w, x) = -409391.47958 + 866.14641*temperature + 1032.69507* Mols KCL$$

Report on score and explain meaning: The score is 0.8552477276095, and is residual sum of squares. This score measures how well the model's predictions align with the actual values. A score of 1 indicates that the model's predictions perfectly match the true values, while a score of 0 means no alignment. Based on the returned score, the Linear Regression model demonstrates relatively good accuracy about 85% matching.

## 4. Use Cross Validation

In [13]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
# Report on their finding and their significance
cross_valid_QL = KFold(n_splits=5, random_state=42, shuffle=True)
cross_scores = cross_val_score(applied_Linear_Regression_model,X_features,Y_label,cv=cross_valid_QL)
mean_of_folds = cross_scores.mean()
stdard_QLoc_diation = cross_scores.std()

folds_score_BR_data_frame = pd.DataFrame({
    'Fold': np.arange(1, len(cross_scores) + 1),
    'Score': cross_scores
})
metrics = pd.DataFrame({
    'Mean' : [mean_of_folds],
    'standard Deviation': [stdard_QLoc_diation]
})
display(folds_score_BR_data_frame)
display(metrics) 

,Fold,Score
0,1,0.861519
1,2,0.827423
2,3,0.871952
3,4,0.881662
4,5,0.856091


,Mean,standard Deviation
0,0.859729,0.018388


Write findings here: The `cross_val_score` function returns five different values (0.861519, 0.827423, 0.871952, 0.881662, 0.856091) corresponding to the five splits. The scores do not vary significantly, as the standard deviation is approximately 0.018. The mean accuracy is about 0.86, indicating that the model has performed quite well during training. After rerun multiple times, those values remain unchanged

## 5. Using Polynomial Regression

In [14]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly_regeresion_form = PolynomialFeatures(degree=2, include_bias= False)
X_features_in_poly_form = poly_regeresion_form.fit_transform(X_features)

applied_Linear_Regression_model.fit(X_features_in_poly_form,Y_label)

coefficientd_values_after_poly_transform = applied_Linear_Regression_model.coef_

intercept_value_after_poly_transform = applied_Linear_Regression_model.intercept_

scores_on_poly_cross_validation = cross_val_score(applied_Linear_Regression_model,X_features_in_poly_form,\
                                                  Y_label,cv=cross_valid_QL)

poly_re_mean = scores_on_poly_cross_validation.mean()
poly_re_std_dev = scores_on_poly_cross_validation.std()
dummi_data_sample_after_poly = poly_regeresion_form.fit_transform(dummi_data_sample)

poly_re_data_fr = pd.DataFrame({
    "predict sample output": [applied_Linear_Regression_model.predict(dummi_data_sample_after_poly)],
    "coefficents" : [coefficientd_values_after_poly_transform.round(5)],
    "intercept": [intercept_value_after_poly_transform.round(5)],
    "scores": [scores_on_poly_cross_validation],
    "mean": [poly_re_mean],
    "standard deviation": [poly_re_std_dev],
})
display(poly_re_data_fr)

# Report on the metrics and output the resultant equation as you did in Part 3.

,predict sample output,coefficents,intercept,scores,mean,standard deviation
0,[132543.4571369136],"[12.0, -0.0, -0.0, 2.0, 0.02857]",0.00002,"[1.0, 1.0, 1.0, 1.0, 1.0]",1.0,0.0


Write the polynomial equation of a slime:
- At first, call $$z = [x_1, x_2, x_1^2 , x_1 x_2,x_2^2]$$

- $$w = [12, 0,  0, 2, 0.02857]$$

- Thus, 
$$h(x) = 2 \times 10^{-5} + 12*temperature + 2*temperature* MolsKCL + 0.02857*(MolsKCL)^2 $$ 
Or
$$h(z) = 2 \times 10^{-5} + 12z_1 + 2z_4 + 0.02857z_5 $$

Report on the score and interpret: The returned results are [1, 1, 1, 1, 1], which indicates that the polynomial regression predictions perfectly match the actual values, with no variance among the 5 folds (std == 0 ), and the mean value is 1. We can also say the polynomial regression has flexibility to fit much more data than the linear model.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [ ]:
# Load the dataset. Then train and evaluate the classification models.
# Imports section
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import KFold
from sklearn.neural_network import MLPClassifier

load_ckd_feature_set = pd.read_csv("ckd_feature_subset.csv")
cross_valid_QL = KFold(n_splits=5, random_state=42, shuffle=True)
X_features_ckd_feature = load_ckd_feature_set.iloc[:,:-1]
Y_label_ckd_feature = load_ckd_feature_set.iloc[:,-1]
display(load_ckd_feature_set.head(15))
display(load_ckd_feature_set.info())


classification_training_models = {
    "logistic_re_model" : LogisticRegression(),
    "support_vactor_mac_modl" : svm.SVC(),
    "K_near_neighbor_model" : KNeighborsClassifier(),
    "Neural Network": MLPClassifier()
}

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
5,0.740260,0.833333,0.265487,0,1,0.355932,1
6,0.870130,0.333333,0.668142,0,1,0.237288,1
7,0.870130,0.500000,0.150442,0,1,0.372881,1
8,0.194805,0.666667,0.380531,0,1,0.305085,1
9,0.753247,0.833333,0.163717,0,1,0.220339,1


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB


None

In [17]:
def Test_Classification_Models (classification_training_models):
    for key, learning_model_training in classification_training_models.items():
        learning_model_score = cross_val_score(learning_model_training,X_features_ckd_feature,\
                                           Y_label_ckd_feature,cv=cross_valid_QL,scoring='accuracy') 
        mean_of_folds_of_model = learning_model_score.mean()
        std_dev_of_models = learning_model_score.std()
Test_Classification_Models(classification_training_models)

/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.1

## Results and Conclusion for Classification Experiments
| Model | Cross Scores | Mean Accuracy | Standard Deviation |
|--------------------------|--------------------------------------------------|---------------|---------------------|
| Logistic Regression | [0.8065, 0.9355, 0.7742, 0.8333, 0.9333] | 0.8566 | 0.0663 |
| Support Vector Machines | [0.9032, 1.0000, 0.8710, 0.9000, 0.9667] | 0.9282 | 0.0476 |
| k-Nearest Neighbors | [0.9355, 1.0000, 0.8710, 0.8667, 0.9667] | 0.9280 | 0.0524 |
| Neural Network | [0.839 0.9678 0.7742 0.87 0.97] | 0.8828 | 0.075164 |

Observations and Conclusion:
- Based on the mean accuracy and standard deviation across 5 folds, Support Vector Machines (SVM) is currently the best model, achieving a mean accuracy of 0.9282 with a standard deviation of 0.0476. K-Nearest Neighbors (KNN) is also a good model, with a mean accuracy of 0.9280 and a standard deviation of 0.0524. In contrast, Logistic Regression shows the lowest performance, with a mean accuracy of only 0.8566 and considerable variability, as evidenced by the accuracy of 0.7742 in the third fold. Although the Neural Network model performs slightly better than Logistic Regression, it remains suboptimal as the optimization has not converged even after the maximum of 200 iterations. Therefore, I need to conduct further experiments to evaluate how changes in parameters may affect the results of the Neural Network model. One more thing about this model is the results keep changing. 

In [ ]:
# Experiments with Neural Network.
#1 change hidden layer size:
classification_training_models ["Neural Network Change Hidden Layer Size to 500"]\
      = MLPClassifier(hidden_layer_sizes= [500,500,500])
#1.1 change hidden layer size 1:
classification_training_models ["Neural Network Change Hidden Layer Size to 10"]\
      = MLPClassifier(hidden_layer_sizes= 10)
#2 change activation:
classification_training_models ["Neural Network Change from relu to tanh"]\
        = MLPClassifier( activation="tanh")
#3 change solver:
classification_training_models ["Neural Network Change adam -> sgd"] \
        = MLPClassifier(solver="sgd")
Test_Classification_Models(classification_training_models)

/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.1

## Results and Conclusion for Neural Network Experiments
| Model | Cross Scores | Mean Accuracy | Standard Deviation |
|--------------------------------------------------|--------------------------------------------------|---------------|---------------------|
| Logistic Regression | [0.8065, 0.9355, 0.7742, 0.8333, 0.9333] | 0.8566 | 0.0663 |
| Support Vector Machines | [0.9032, 1.0000, 0.8709, 0.9000, 0.9667] | 0.9282 | 0.0476 |
| K-Nearest Neighbors | [0.9355, 1.0000, 0.8709, 0.8667, 0.9667] | 0.9280 | 0.0524 |
| Neural Network | [0.8387, 0.9677, 0.7742, 0.8667, 0.9667] | 0.8828 | 0.0752 |
| Neural Network (Hidden Layer Size = [500,500...]) | [0.96774194 1.         0.90322581 1.         0.96666667] | 0.9676 | 0.03534 |
| Neural Network (Hidden Layer Size = 10) | [0.6774, 0.9355, 0.7742, 0.8333, 0.9000] | 0.8241 | 0.0919 |
| Neural Network (Activation: ReLU to Tanh) | [0.8387, 0.9355, 0.8065, 0.9000, 0.9667] | 0.8895 | 0.0594 |
| Neural Network (Optimizer: Adam to SGD) | [0.7419, 0.7097, 0.6774, 0.8000, 0.8000] | 0.7458 | 0.0487 |

In summary, I conducted experiments on Layer Size, Activation functions, and Solvers. First, I examined how changes in Layer Size affect the results. It appears that using a relatively large Layer Size (around 500 or more, and two or more layers) significantly enhances the performance of the Neural Network, achieving a mean accuracy of 0.9676 compared to 0.8828, and the standard deviation of 0.03534 vs 0.0752. Conversely, when the Layer Size is small, the performance either decreases or remains unchanged. Next, I investigated whether changing the activation function in the hidden layers had a significant impact. As shown in the table, switching from ReLU to Tanh did not result in any meaningful change in performance. Lastly, I observed that changing the solver from Adam to SGD led to a significant decrease in results, with accuracy dropping from 0.8828 to 0.7458. 

As a conclusion, solver and Layer Size have strong impact on performance of Neural Network. 